#usage: Rscript --vanilla TCGA_gene_depths_pairs_from_VCF.R TP53

args<-commandArgs(TRUE)
.libPaths()
# test whether both arguments present
if (length(args)!=2) {
  stop("Usage example: Rscript --vanilla TCGA_gene_depths_pairs_from_MAF.R TP53 ClinVarYes", call.=FALSE)
} else if (length(args)==2) {
  # default output file
  GENE <- toString(args[1])
  CLINVAR <- toString(args[2])
}

#load libraries
.libPaths(c( .libPaths(), "/home/mayo/m187735/R", "/usr/local/biotools/rpackages/R-3.5.2-latest",  "/usr/local/biotools/r/R-3.5.2/lib64/R/library"))
library(plyr)
library(dplyr)
library(readr)
library(tidyr)
library(stringr)
library(gtools)
setTimeLimit(cpu = Inf, elapsed = Inf, transient = FALSE)

############################################
#INDIVIDUAL GENE ANALYSIS: EACH ROW IS A DETECTED SOMATIC MUTATION FROM ONE PARTICULAR PATIENT#
############################################

CANCER_LIST <- c('KIRP','GBM','BRCA','LUAD','UCEC','COAD','PRAD','KIRC','HNSC')


for (CANCER_NAME in CANCER_LIST){
  
  #Extract mutations in particular genes from protected MAF
  system(paste0("awk  '$1 == \"Hugo_Symbol\" || $1 == \"",GENE,"\" &&  ($12 == \"A\" || $12 == \"G\" || $12 == \"T\" || $12 == \"C\") && ($13 == \"A\" || $13 == \"G\" || $13 == \"T\" || $13 == \"C\")' /research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/MAFs/protected_hg38/",CANCER_NAME,"_protected.maf > /research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/processed_data/MAFs/protected_hg38/",CANCER_NAME,"_",GENE,"_protected.tmp"))
  
  #Extract mutations in particular genes from controlled MAF
  # system(paste0("awk  '$1 == \"Tumor_Sample_Barcode_short\" || $2==\"",DRIVER_GENES[1],"\" && $3 >=",DRIVER_GENES[2]," && $3 <=",DRIVER_GENES[3]," &&  ($8 == \"A\" || $8 == \"G\" || $8 == \"T\" || $8 == \"C\") && ($9 == \"A\" || $9 == \"G\" || $9 == \"T\" || $9 == \"C\")' /research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/MAFs/controlled_MAFs_per_cancer/TCGA-",CANCER_NAME,"_from_controlled_MAF_liftover_hg38.maf > /research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/MAFs/controlled_MAFs_per_cancer/",CANCER_NAME,"_",GENE,"_controlled.maf"))
  
  cdriver_tumor_mutations_protected <- read.table(paste0("/research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/processed_data/MAFs/protected_hg38/",CANCER_NAME,"_",GENE,"_protected.tmp"),header=TRUE,sep='\t',quote="")
  cdriver_tumor_mutations_protected <- cdriver_tumor_mutations_protected[,c(1,5,6,7,9,10,12,13,16,17,18,19,35,36,40,41,42)]
  
  #cdriver_tumor_mutations <- read.table(paste0("/research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/MAFs/controlled_MAFs_per_cancer/",CANCER_NAME,"_",GENE,"_controlled.maf"),header=TRUE,sep='\t',quote="")
  
  
  
  #Extract mutations in particular genes from somatic MAFs
  # system(paste0("awk  '$1 == \"Hugo_Symbol\" || $1 == \"",GENE,"\" &&  ($12 == \"A\" || $12 == \"G\" || $12 == \"T\" || $12 == \"C\") && ($13 == \"A\" || $13 == \"G\" || $13 == \"T\" || $13 == \"C\")' /research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/MAFs/",CANCER_NAME,"_somatic.maf > /research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/MAFs/",CANCER_NAME,"_",GENE,"_somatic.tmp"))
  
  #  cdriver_tumor_mutations_somatic <- read.table(paste0("/research/bsi/projects/PI/tertiary/Asmann_Yan_wangy3/s212975.Wickland_Immunomics/processing/TCGA/MAFs/",CANCER_NAME,"_",GENE,"_somatic.tmp"),header=TRUE,sep='\t',quote="")
  #  cdriver_tumor_mutations_somatic <- cdriver_tumor_mutations_somatic[-c(18,19,44,45)]
  
  #combine the two, i.e. get the germline info from the protected and add to the somatic
  #  cdriver_tumor_mutations <- merge(cdriver_tumor_mutations_somatic,cdriver_tumor_mutations_protected, by=c('Chromosome','Start_Position','End_Position','Tumor_Sample_Barcode','Matched_Norm_Sample_Barcode','Tumor_Seq_Allele1','Tumor_Seq_Allele2'))
  
  cdriver_tumor_mutations <- cdriver_tumor_mutations_protected
  names(cdriver_tumor_mutations)[7] <- 'Tumor_Seq_Ref'
  names(cdriver_tumor_mutations)[8] <- 'Tumor_Seq_Alt'
  
  #add additional columns: total AD, alt allele concentration (from AD)
  cdriver_tumor_mutations$alt_allele_concent_MAF <- as.numeric(cdriver_tumor_mutations$t_alt_count) / as.numeric(cdriver_tumor_mutations$t_depth)
  
  
  #3. Match with TCGA metadata
  
  #shorten tumor bacode / submitter_id
  cdriver_tumor_mutations$submitter_id_long <- sub("^([^-]*-[^-]*-[^-]*-[^-]*).*", "\\1",cdriver_tumor_mutations$Tumor_Sample_Barcode)
  cdriver_tumor_mutations$submitter_id <- sub("^([^-]*-[^-]*-[^-]*).*", "\\1",cdriver_tumor_mutations$Tumor_Sample_Barcode)
  
  
  #4. Add exome-bams-metadata and exome bam paths (from overall read depths data generated by previous script)
  
  #Get metadata for all bams; duplicates fine b/c some patients will have multiple mutations 
  exome_bams_metadata <- read.table(paste('/home/mayo/m187735/s212975.Wickland_Immunomics/TCGA_metadata/exome_bams/',CANCER_NAME,'_bams_metadata.txt',sep=''),header=TRUE)
  exome_bams_metadata$submitter_id <- sub("^([^-]*-[^-]*-[^-]*-[^-]*).*", "\\1",exome_bams_metadata$cases)
  
  #only interested in primary solid tumor
  exome_bams_metadata  <- subset(exome_bams_metadata, (tissue.definition=='Primary solid Tumor'))
  
  colnames(exome_bams_metadata)[colnames(exome_bams_metadata) == 'submitter_id'] <- 'submitter_id_long'
  cdriver <- merge(cdriver_tumor_mutations,exome_bams_metadata,by='submitter_id_long',all.x=TRUE)
  rm(exome_bams_metadata)
  
  #subset only the primary solid tumor bams --i.e.the indivs from VCFs for whom we have bams
  cdriver <- subset(cdriver, (tissue.definition=='Primary solid Tumor'))
  
  
  #Get overall read depths data
  combined_data <- read.table(paste0('/home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/processed_data/overall_depths/',CANCER_NAME,'_overall_depths_etc.txt'),header=TRUE)
  
  #perhaps better to combine based on exome bam filename, rather than on submitter_id????
  combined_data$exome_filename_without_path <- sub(".*\\/", "\\1",combined_data$bam_full_path_exome) #delete path
  
  #combine combined_data (containing overall read depths) and cdriver data; some patients with VCFs may not have bams, and vice versa; combined based on exome file name, rather than submitter_id (less ambiguity)
  cdriver <- merge(cdriver, combined_data, by.x='file_name',by.y='exome_filename_without_path')  #note that no. of mutations goes down v slightly here....investigate later
  
  #remove extraneous columns
  names(cdriver)[2] <- 'submitter_id_long'
  #  names(cdriver)[124] <- 'submitter_id'
  names(cdriver)[21] <- 'submitter_id'
  #add col for number of mutations per sample
  cdriver <- add_count(cdriver, submitter_id)
  names(cdriver)[ncol(cdriver)] <- 'mutation_count_for_patient'
  
  
  #4.5 subset only the loci designated as Pathogenic in ClinVar file, if desired
  if (CLINVAR == 'ClinVarYes'){
  
  #extract gene from ClinVar VCF
      system(paste0("cat /home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/ClinVar/GRCh38_latest_clinvar.vcf | awk '/#/ || length($4) == 1 && length($5) ==1' | grep -w ",GENE," | awk '$8 ~ \"Pathogenic\" || $8 ~ \"Likely pathogenic\"' > /home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/ClinVar/",GENE,"_",CANCER_NAME,".tmp"))
  
  #add header and append 'chr' to chromosome col
      ClinVar_Mutations <- read.table(paste0("/home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/ClinVar/",GENE,"_",CANCER_NAME,".tmp"), sep='\t',header=FALSE,quote="")
      colnames(ClinVar_Mutations) <- c('Chromosome','Start_Position','ID','Tumor_Seq_Ref','Tumor_Seq_Alt','QUAL','FILTER','INFO')
      ClinVar_Mutations <- ClinVar_Mutations[,c(1,2,4,5)]
      ClinVar_Mutations$Chromosome <- paste0("chr",ClinVar_Mutations$Chromosome)
  
      cdriver <- merge(cdriver, ClinVar_Mutations, by=c('Chromosome','Start_Position','Tumor_Seq_Ref','Tumor_Seq_Alt'),all.y=TRUE)
  }
  #are any patients represented by more than one bam??????????????
  
  
  #5 Extract read depths from VCF files
  
  
  
  
  #5. Read depth calculations: Count number of mapped reads of each cdriver mutated position in each individual
  variant_read_depth_bam_VIEW_col <- ncol(cdriver)+1
  variant_read_depth_bam_MPILEUP_col <- ncol(cdriver)+2
  tumor_ref_base_count_bam_col <- ncol(cdriver)+3
  tumor_alt_base_count_bam_col <- ncol(cdriver)+4
  tumor_ref_avg_base_quality_bam_col <- ncol(cdriver)+5
  tumor_alt_avg_base_quality_bam_col <- ncol(cdriver)+6
  
  for (i in 1:nrow(cdriver)){ 
    print(paste0(CANCER_NAME,": ",i,"/",nrow(cdriver)))
    #total_read_depth <- numeric(0) #initialize object to store depths for particular case ID / row
    #row_position_read_depth <- numeric(0) #initialize object to store depths for particular case ID / row
    #output NA if index file does not exist
    variant_read_depth_bam_VIEW <- system(paste("if [ -f ",sub(".bam",".bai",cdriver$bam_full_path_exome[i])," ]; then /research/bsi/tools/biotools/samtools/1.9/miniconda/bin/samtools view -c -F4 -F1024 -q 20",cdriver$bam_full_path_exome[i],paste(cdriver$Chromosome[i],":",cdriver$Start_Position[i],"-",cdriver$End_Position[i],"; else echo NA; fi",sep='')),intern=TRUE,ignore.stderr=TRUE)
    cdriver[i,variant_read_depth_bam_VIEW_col] <- variant_read_depth_bam_VIEW
    
    pileup <- system(paste("if [ -f ",sub(".bam",".bai",cdriver$bam_full_path_exome[i])," ]; then /research/bsi/tools/biotools/samtools/1.9/miniconda/bin/samtools mpileup ",cdriver$bam_full_path_exome[i],"--min-MQ 20 --min-BQ 0 --count-orphans --output-MQ --region ",paste(cdriver$Chromosome[i],":",cdriver$Start_Position[i],"-",cdriver$End_Position[i],"; else echo NA; fi",sep='')),intern=TRUE,ignore.stderr=TRUE)
    
    if (length(pileup)!=0 && pileup!= 'NA'){
      variant_read_depth_bam_MPILEUP <- strsplit(pileup,'\t')[[1]][4] 
      bases <- strsplit(pileup,'\t')[[1]][5]
      bases <- gsub("[^tcgaTCGA+]+","",bases) #remove $ character from list of bases; this simply; don't need to do this for the qualities b/c no. qual values is already equal to no. bases
      
      #upper case bases are positive strand, while lower-case bases are negative strand; MAF file lists only positive, so must reverse the lower-case bases in pileup ??????????????????????????????????????
      bases <- gsub("g","G", bases)
      bases <- gsub("c","C", bases)
      bases <- gsub("a","A", bases)
      bases <- gsub("t","T", bases)
      
      
      tumor_ref_base_count_bam <- str_count(bases,as.character(cdriver[i,]$Tumor_Seq_Ref)) #count number of reads with ref base in the pileup (convert all to uppercase)
      tumor_alt_base_count_bam <- str_count(bases,as.character(cdriver[i,]$Tumor_Seq_Alt)) #count number of reads with alt base in the pileup (convert all to uppercase)
      
      cdriver[i,variant_read_depth_bam_MPILEUP_col] <- variant_read_depth_bam_MPILEUP
      cdriver[i, tumor_ref_base_count_bam_col] <- tumor_ref_base_count_bam
      cdriver[i, tumor_alt_base_count_bam_col] <- tumor_alt_base_count_bam
      
      tumor_ref_indices <- unlist(str_locate_all(bases,as.character(cdriver[i,]$Tumor_Seq_Ref)))[1:tumor_ref_base_count_bam] #identify index positions of REF allele in the pileup; then get their corresponding quality scores; unlisting must be chopped in half b/c entire vector repeated once
      tumor_alt_indices <- unlist(str_locate_all(bases,as.character(cdriver[i,]$Tumor_Seq_Alt)))[1:tumor_alt_base_count_bam] #identify index positions of REF allele in the pileup; then get their corresponding quality scores; unlisting must be chopped in half b/c entire vector repeated once
      
      #    read_mapping_quality <- asc(strsplit(pileup,'\t')[[1]][7])-33
      #    avg_read_mapping_quality_ref_bam <- round(mean(read_mapping_quality[ref_indices]),digits=2)
      #    avg_read_mapping_quality_alt_bam <- round(mean(read_mapping_quality[alt_indices]),digits=2) #problem: insertion sequences; see http://samtools.sourceforge.net/pileup.shtml; so for now, just don't do mapping quality
      
      
      base_quality <- asc(strsplit(pileup,'\t')[[1]][6])-33
      tumor_ref_avg_base_quality_bam <- round(mean(base_quality[tumor_ref_indices]),digits=2)
      tumor_alt_avg_base_quality_bam <- round(mean(base_quality[tumor_alt_indices]),digits=2)
      
      
      
      #  cdriver[i,avg_read_mapping_quality_ref_bam_col] <- avg_read_mapping_quality_ref_bam
      #  cdriver[i,avg_read_mapping_quality_alt_bam_col] <- avg_read_mapping_quality_alt_bam
      
      cdriver[i,tumor_ref_avg_base_quality_bam_col] <-  tumor_ref_avg_base_quality_bam
      cdriver[i,tumor_alt_avg_base_quality_bam_col] <-  tumor_alt_avg_base_quality_bam
      
    }
    
    else if (length(pileup)==0 || pileup=='NA'){
      cdriver[i,variant_read_depth_bam_MPILEUP_col] <- NA
      cdriver[i,tumor_ref_base_count_bam_col] <- NA
      cdriver[i,tumor_alt_base_count_bam_col] <- NA
      
      
      #cdriver[i,avg_read_mapping_quality_ref_bam_col] <- NA
      #cdriver[i,avg_read_mapping_quality_alt_bam_col] <- NA
      cdriver[i,tumor_ref_avg_base_quality_bam_col] <- NA
      cdriver[i,tumor_alt_avg_base_quality_bam_col] <- NA
      
    }
    
    #total_read_depth <- system(paste("/research/bsi/tools/biotools/samtools/1.9/miniconda/bin/samtools view -c -F4",cdriver$file_name_exome_bams[i]),intern=TRUE)
    #cdriver[i,total_read_depth_col] <- total_read_depth
    i <- i + 1 #increase row
  }
  
  colnames(cdriver)[variant_read_depth_bam_VIEW_col] <- 'variant_read_depth_bam_VIEW'
  colnames(cdriver)[variant_read_depth_bam_MPILEUP_col] <- 'variant_read_depth_bam_MPILEUP'
  colnames(cdriver)[tumor_ref_base_count_bam_col] <- 'tumor_ref_base_count_bam'
  colnames(cdriver)[tumor_alt_base_count_bam_col] <- 'tumor_alt_base_count_bam'
  #colnames(cdriver)[avg_read_mapping_quality_ref_bam_col] <- 'avg_read_mapping_quality_ref_bam'
  #colnames(cdriver)[avg_read_mapping_quality_alt_bam_col] <- 'avg_read_mapping_quality_alt_bam'
  colnames(cdriver)[tumor_ref_avg_base_quality_bam_col] <- 'tumor_ref_avg_base_quality_bam'
  colnames(cdriver)[tumor_alt_avg_base_quality_bam_col] <- 'tumor_alt_avg_base_quality_bam'
  
  #add additional column: alt allele concentration from bams
  # cdriver$alt_allele_concent_bam <- as.numeric(cdriver$alt_base_count_bam) / as.numeric(cdriver$variant_read_depth_bam_MPILEUP)  
  
  #remove extra columns 
  cdriver <- cdriver[,c("submitter_id","Chromosome","Start_Position","End_Position","Tumor_Seq_Ref","Tumor_Seq_Alt","Match_Norm_Seq_Allele1","Match_Norm_Seq_Allele2","Hugo_Symbol","Variant_Classification","Variant_Type","HGVSc","HGVSp","t_depth","t_ref_count","t_alt_count","alt_allele_concent_MAF","mutation_count_for_patient","variant_read_depth_bam_MPILEUP","tumor_ref_base_count_bam","tumor_alt_base_count_bam","tumor_ref_avg_base_quality_bam","tumor_alt_avg_base_quality_bam","race","center_exome","Source_Site_exome","Cancer","TCGA_Subtype","read_depth_exome","read_depth_RNAseq","bam_full_path_exome","bam_full_path_RNAseq")]
  
  #identify patient + locus pairs that have no entry in MAF
  partially_not_in_MAF <- cdriver[cdriver$Start_Position %in% subset(cdriver, (is.na(submitter_id)))$Start_Position,]
  fully_not_in_MAF <- subset(partially_not_in_MAF, (is.na(submitter_id)))
  try(fully_not_in_MAF$`SNP in at least 1 case in MAF?` <- 'No',silent=TRUE)
  
  #identify patient + locus pairs that have entry in MAF
  partially_in_MAF <- cdriver[cdriver$Start_Position %in% subset(cdriver, (!is.na(submitter_id)))$Start_Position,]
  fully_in_MAF <- subset(partially_in_MAF, (!is.na(submitter_id)))
  fully_in_MAF$`SNP in at least 1 case in MAF?` <- 'Yes'
  
  #combine
  cdriver <- rbind(fully_in_MAF, fully_not_in_MAF)
  
  cdriver$Cancer <- CANCER_NAME
  
  #save dataset
  if (CLINVAR == 'ClinVarNo'){
      system(paste0('mkdir -p /home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/processed_data/gene_depths_pairs_from_MAF/',GENE,'/'))
      system(paste0('rm /home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/processed_data/gene_depths_pairs_from_MAF/',GENE,'/',CANCER_NAME,'_',GENE,'_depths_etc.txt'))
      write.table(x=cdriver,file=paste0('/home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/processed_data/gene_depths_pairs_from_MAF/',GENE,'/',CANCER_NAME,'_',GENE,'_depths_etc.txt'),sep='\t',row.names=FALSE)
  }
  else if (CLINVAR == 'ClinVarYes'){
    system(paste0('mkdir -p /home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/processed_data/gene_depths_pairs_from_MAF/',GENE,'/'))
    system(paste0('rm /home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/processed_data/gene_depths_pairs_from_MAF/',GENE,'/',CANCER_NAME,'_',GENE,'_depths_etc_ClinVar.txt'))
    write.table(x=cdriver,file=paste0('/home/mayo/m187735/s212975.Wickland_Immunomics/processing/TCGA/processed_data/gene_depths_pairs_from_MAF/',GENE,'/',CANCER_NAME,'_',GENE,'_depths_etc_ClinVar.txt'),sep='\t',row.names=FALSE)
  }
}  
